apis/v1alpha2/shared_types.go

```diff
@@ -399,3 +399,65 @@ type AnnotationKey string
 // +kubebuilder:validation:MinLength=0
 // +kubebuilder:validation:MaxLength=4096
 type AnnotationValue string
+
+// AddressRouteMatch defines NetworkAddress rules for inbound traffic according to
+// source and/or destination address of a packet or connection.
+// Multiple match fields are ANDed together, i.e. the match will
+// evaluate to true only if all conditions are satisfied.
+// At least one match field is required.
+type AddressRouteMatch struct {
+	// SourceAddresses indicates the originating (source) network
+	// addresses which are valid for routing traffic.
+	//
+	// Support: Core
+	//
+	// +optional
+	SourceAddress *NetworkAddress `json:"sourceAddresses"`
+
+	// DestinationAddresses indicates the destination network addresses
+	// which are valid for routing traffic.
+	//
+	// Support: Core
+	//
+	// +optional
+	DestinationAddress *NetworkAddress `json:"destinationAddresses"`
```

Comment:
@shaneutt @jpeach I wanted to follow up a bit on why I was hesitant about this change. As I understand it, we've gone from separate lists of source and dest addresses, to lists of source/dest address pairs. That is fine as long as users are only trying to match on source address or dest address, but gets frustrating if they're trying to match on both.

For example, previously the following would have been possible:
```yaml
- source:
  - 10.1.2.3
  - 10.1.2.4
- dest:
  - 10.2.3.4
  - 10.2.3.5
 ```

To represent that now, we end up with an exponential increase in config:
```yaml
- source: 10.1.2.3
  dest:   10.2.3.4
- source: 10.1.2.4
  dest:   10.2.3.4
- source: 10.1.2.3
  dest:   10.2.3.5
- source: 10.1.2.4
  dest:   10.2.3.5
```

Both examples are actually simplified versions of the config that would actually be needed, but hopefully that helps illustrate the concern in terms of exponential growth when both source and dest matching need to be combined.

A potential solution would be to make both SourceAddress and DestinationAddress lists here or say that we can _mostly_ solve this with CIDRs later, but neither of those options feel great.

Reply:
Yeh, the first example is definitely cleaner. I am not sure how realistic this example is (do people really need to deal with routing lots of individual addresses?), but we should probably still prefer the cleaner syntax.